### DISCUSSION TITLE
Next 13 new "App Router" in combination with progressive enhancement pattern

### DISCUSSION THREAD

**Discussion by eugeniosegala** [vercel/next.js#51184] [Help] [answered]
### Summary

Hey,

I'm looking at migrating an application to the new "App Router" paradigm released in Next 13.

Currently, one of the key requirements for my application is to be ["Progressively enhanced"](https://en.wikipedia.org/wiki/Progressive_enhancement).

A "Progressively enhanced" application is an application which can work when JavaScript is disabled or not loaded (at least for the basic functionalities), then when JavaScript is enabled I'm allowed to introduce more functionality and effectively enhance the experience.

I'm currently able to build pages which are progressively enhanced using `getServerSideProps`.

Basically what I'm doing is the following:

```

// Client side code
<form onSubmit={handleSubmit((data) => onSubmit(data))} method="POST">
// rest of the code...

// Server side code
export const getServerSideProps = nextServerStore.getServerSideProps(
  (store) =>
    async ({ req }: { req: ExtendedNextApiRequest }) => {
      let todoId = null;

      const postTodoValue = req.body?.todoId;

      // dispatching Redux actions server side
      if (req.method === "POST" && postTodoValue) {
        todoId = Number(postTodoValue);

        // init REST getTodos SSR query
        store.dispatch(getTodo.initiate(todoId));

        // rest endpoints
        await Promise.all(
          store.dispatch(restAPI.util.getRunningQueriesThunk())
        );
      }

      return {
        props: {
          todoId,
        },
      };
    }
);
```

As you can see, if JavaScript is disabled, the html form will fallback on `method="POST"` and trigger a page submission. Then, I'm reading the data on `getServerSideProps` coming from the form POST: `req.body?.todoId` which I will use to perform the HTTP request (which I would usually perform client side) and get the data I need.

To reduce the amount of code duplicated, I'm using [Redux Toolkit RTK](https://redux-toolkit.js.org/rtk-query/usage/server-side-rendering), so that's why you see the `store.dispatch` etc...

I was wondering, how can I replicate this behaviour with the new app routing approach? 

I could not find a way to read `req.body?.todoId` and similar.

Any ideas?

**Comment by devjiwonchoi** [comment, marked as answer]
`Server Actions` is exactly what you are looking for.

> [Server Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#server-actions) are an alpha feature in Next.js, built on top of React [Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#actions). They enable server-side data mutations, reduced client-side JavaScript, and progressively enhanced forms.

You need some work around for Client Components though.

```js
// app/actions.ts
'use server'
 
export async function myAction() {
  // ...
}
```

```js
// client-side form component
'use client'
 
import { myAction } from './actions'
 
export default function ClientComponent() {
  return (
    <form action={myAction}>
      <button type="submit">Add to Cart</button>
    </form>
  )
}
```

Note that the `Server Actions` is still in the `alpha` stage, and need to set the `experimental` config on `next.config.js`.

```
// next.config.js

module.exports = {
  experimental: {
    serverActions: true,
  },
}
```

## Reference
- [Server Actions](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions)
- [Server Actions with Client Components](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#with-client-components)

**Reply by eugeniosegala** [reply]
Thank you! I did not know about this feature :)

Do you know if there are any plans to make the `req.body` available to the "Server Actions" so it will be possible to read data that are sent from html forms?

**Reply by devjiwonchoi** [reply]
Not at the moment, but I'm going to open a PR for that feature. I've looked at the current source and it is only designed for the `Page Router` assuming that the features for stabilizing the Server Actions are prioritized, and seems like it'll need some time to add the feature for managing the `Request` and `Response` as you've stated. I'll let you know when it happens.

**Reply by karlguillotte** [reply]
Do you have a link available to that PR? It is a blocker to our migration to the "app directory".
